python/mxnet/optimizer.py

```diff
@@ -143,8 +143,8 @@ def _update_count(self, index):
         """
         if index not in self._index_update_count:
             self._index_update_count[index] = self.begin_num_update
-        self._index_update_count[index] += 1
-        self.num_update = max(self._index_update_count[index], self.num_update)
```

Comment:
the problem of original code is that the first grad and others get different value of the num_update in each iteration.  Maybe you can debug it.